### PR TITLE
[vcpkg] Fix the case of current_path() before use on Windows.

### LINKED ISSUE #13105: [zlib] AppVeyor build failure
**Host Environment**
- OS: AppVeyor Visual Studio 2017 Image
- Compiler: Visual Studio 2017

**To Reproduce**
Execute `./vcpkg install zlib` on the AppVeyor Visual Studio 2017 image.

**Failure logs** 
```
-- Downloading http://www.zlib.net/zlib-1.2.11.tar.gz...
-- Extracting source C:/tools/vcpkg/downloads/zlib1211.tar.gz
-- Applying patch cmake_dont_build_more_than_needed.patch
-- Applying patch 0001-Prevent-invalid-inclusions-when-HAVE_-is-set-to-0.patch
-- Applying patch add_debug_postfix_on_mingw.patch
-- Using source at C:/tools/vcpkg/buildtrees/zlib/src/1.2.11-63309e48e5.clean
-- Configuring x86-windows
-- Building x86-windows-dbg
-- Building x86-windows-rel
-- Downloading https://repo.msys2.org/mingw/i686/mingw-w64-i686-pkg-config-0.29.2-1-any.pkg.tar.xz...
-- Downloading https://repo.msys2.org/mingw/i686/mingw-w64-i686-libwinpthread-git-8.0.0.5906.c9a21571-1-any.pkg.tar.zst...
-- Fixing pkgconfig file: C:/tools/vcpkg/packages/zlib_x86-windows/lib/pkgconfig/zlib.pc
-- Fixing pkgconfig file: C:/tools/vcpkg/packages/zlib_x86-windows/debug/lib/pkgconfig/zlib.pc
-- CHECK_LIB_zlibd_DEBUG:CHECK_LIB_zlibd_DEBUG-NOTFOUND
CMake Error at scripts/cmake/vcpkg_fixup_pkgconfig.cmake:232 (message):
  Library "zlibd" was not found! If it is a system library use the
  SYSTEM_LIBRARIES parameter for the vcpkg_fixup_pkgconfig call! Otherwise,
  correct the *.pc file
Call Stack (most recent call first):
  scripts/cmake/vcpkg_fixup_pkgconfig.cmake:329 (vcpkg_fixup_pkgconfig_check_files)
  ports/zlib/portfile.cmake:48 (vcpkg_fixup_pkgconfig)
  scripts/ports.cmake:79 (include)
Error: Building package zlib:x86-windows failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: zlib:x86-windows
  Vcpkg version: 2020.06.15-nohash
Additionally, attach any relevant sections from the log files above.
Failed to build static libs
```

**Additional context**
This works on my local machine with Visual Studio 2019. Here is an example [appveyor.yml file](https://github.com/Hoikas/PlasmaPrefix/blob/681287437d022a8fb9e4560c3a41a8ab2cb71b54/appveyor.yml) and here is an example of a [failing build](https://ci.appveyor.com/project/Hoikas/plasmaprefix/builds/34810402).

### PR DESCRIPTION
**Pull request by BillyONeal**
This is a better solution than that I tried in https://github.com/microsoft/vcpkg/pull/13144 -- rather than trying to cannoicalize the path completely, which would destroy symlinks etc; this form  calls FindFirstFile on each path element to get the real case from the filesystem.

Fixes #13105.

(Note the wrong case in the CWD in the prompt, but the correct case in all the output:)

```
PS C:\DeV\vcPKG> .\vcpkg.exe install curl
Computing installation plan...
The following packages will be built and installed:
    curl[core,non-http,schannel,ssl,sspi,winssl]:x86-windows
  * zlib[core]:x86-windows
Additional packages (*) will be modified to complete this operation.
Detecting compiler hash for triplet x86-windows...
Starting package 1/2: zlib:x86-windows
Building package zlib[core]:x86-windows...
Using cached binary package: C:\Users\billy\AppData\Local\vcpkg/archives\c6\c61dd1bcc23348934c55f4ced77f1e4b0f353394.zipBuilding package zlib[core]:x86-windows... done
Installing package zlib[core]:x86-windows...
Installing package zlib[core]:x86-windows... done
Elapsed time for package zlib:x86-windows: 62.26 ms
Starting package 2/2: curl:x86-windows
Building package curl[core,non-http,schannel,ssl,sspi,winssl]:x86-windows...
Could not locate cached archive: C:\Users\billy\AppData\Local\vcpkg/archives\9f\9fa16d473c9539e9ea7ab3922eff46618f3a4c4b.zip
-- Downloading https://github.com/curl/curl/archive/9d954e49bce3706a9a2efb119ecd05767f0f2a9e.tar.gz...
-- Extracting source C:/Dev/vcpkg/downloads/curl-curl-9d954e49bce3706a9a2efb119ecd05767f0f2a9e.tar.gz
-- Applying patch 0002_fix_uwp.patch
-- Applying patch 0004_nghttp2_staticlib.patch
-- Applying patch 0005_remove_imp_suffix.patch
-- Applying patch 0006_fix_tool_depends.patch
-- Applying patch 0007_disable_tool_export_curl_target.patch
-- Applying patch 0009_fix_openssl_config.patch
-- Applying patch 0010_fix_othertests_cmake.patch
-- Applying patch 0011_fix_static_build.patch
-- Using source at C:/Dev/vcpkg/buildtrees/curl/src/767f0f2a9e-91d24adee1.clean
-- Configuring x86-windows
-- Building x86-windows-dbg
-- Building x86-windows-rel
-- Installing: C:/Dev/vcpkg/packages/curl_x86-windows/share/curl/curl-config
-- Installing: C:/Dev/vcpkg/packages/curl_x86-windows/share/curl/vcpkg-cmake-wrapper.cmake
-- Installing: C:/Dev/vcpkg/packages/curl_x86-windows/share/curl/copyright
-- Performing post-build validation
-- Performing post-build validation done
Stored binary cache: C:\Users\billy\AppData\Local\vcpkg/archives\9f\9fa16d473c9539e9ea7ab3922eff46618f3a4c4b.zip
Building package curl[core,non-http,schannel,ssl,sspi,winssl]:x86-windows... done
Installing package curl[core,non-http,schannel,ssl,sspi,winssl]:x86-windows...
Installing package curl[core,non-http,schannel,ssl,sspi,winssl]:x86-windows... done
Elapsed time for package curl:x86-windows: 1.299 min

Total elapsed time: 1.33 min

The package curl:x86-windows provides CMake targets:

    find_package(CURL CONFIG REQUIRED)
    target_link_libraries(main PRIVATE CURL::libcurl)

PS C:\DeV\vcPKG>
```

### CODE REVIEW

**Review by ras0219 — APPROVED** [review]

File: toolsrc/src/vcpkg/base/files.cpp [hunk 1]
```diff
@@ -1197,4 +1267,105 @@ namespace vcpkg::Files
...
+            first = std::find_if_not(next_slash, last, is_slash);
+            next_slash = std::find_if(first, last, is_slash);
+            in_progress.append(first, next_slash);
+            first = std::find_if_not(next_slash, last, is_slash);
+            if (first != next_slash)
+            {
+                in_progress.push_back(L'\\');
+            }
+        }
+        else if (last - first >= 1 && is_slash(first[0]))
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply toolsrc/src/vcpkg/base/files.cpp:1315]: Yes; that's consistent with what the STL does. Normally the number of `\\`s you have is not relevant; they together get treated as one directory-separator. Only the special case of 2 `\\`s at the start of the path do we interpret it as a UNC path. See https://github.com/microsoft/STL/blob/5f3e91211ab25a2f493edf5f7d28520ac6b81d28/stl/inc/filesystem#L432

As for the underlying OS, the system usually treats multiple `\`s as one `\` but not always:
```
Microsoft Windows [Version 10.0.19041.508]
(c) 2020 Microsoft Corporation. All rights reserved.

C:\Users\billy>dir \Windows\?ystem32
 Volume in drive C is OS
 Volume Serial Number is E433-5610

 Directory of C:\Windows

09/15/2020  02:19 AM    <DIR>          System32
               0 File(s)              0 bytes
               1 Dir(s)  1,130,906,918,912 bytes free

C:\Users\billy>dir \\Windows\?ystem32
The network path was not found.

C:\Users\billy>dir \\\Windows\?ystem32
The specified path is invalid.

C:\Users\billy>dir \Windows\\\\\?ystem32
 Volume in drive C is OS
 Volume Serial Number is E433-5610

 Directory of C:\Windows

09/15/2020  02:19 AM    <DIR>          System32
               0 File(s)              0 bytes
               1 Dir(s)  1,130,906,255,360 bytes free

C:\Users\billy>
```

File: toolsrc/src/vcpkg/base/files.cpp [hunk 2]
```diff
@@ -1197,4 +1267,105 @@ namespace vcpkg::Files
 #endif // ^^^ windows
 #endif // ^^^ std::experimental::filesystem
     }
+
+#ifdef _WIN32
+    fs::path win32_fix_path_case(const fs::path& source)
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply toolsrc/src/vcpkg/base/files.cpp:1272]: Relative paths are possible if the user supplies one on the command line. It's not difficult to support them so this component should, even though most of the time the input will be absolute.

[the remark this answers is not in the document]
Comment (BillyONeal) [reply toolsrc/src/vcpkg/base/files.cpp:1272]: Can you push a commit fixing up those cases into this branch?

[the remark this answers is not in the document]
Comment (ras0219-msft) [reply toolsrc/src/vcpkg/base/files.cpp:1272]: In those cases, we would prepend the "relative-to" path (via original_cwd) before washing it through this function.

File: toolsrc/src/vcpkg/base/files.cpp [hunk 3]
```diff
@@ -5,19 +5,93 @@
 #include <vcpkg/base/system.process.h>
 #include <vcpkg/base/util.h>
 
-#if !defined(_WIN32)
+#if defined(_WIN32)
+#include <Windows.h>
+#else // ^^^ _WIN32 // !_WIN32 vvv
```
Comment (ras0219) [inline toolsrc/src/vcpkg/base/files.cpp:10]:
```suggestion
#if defined(_WIN32)
#include <vcpkg/base/system_headers.h>
#else // ^^^ _WIN32 // !_WIN32 vvv
```

**Review by ras0219 — APPROVED** [review]

**Pull request merged**